// web3x/shh/shh.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const subscriptions_1 = require("../subscriptions");
class Shh {
    constructor(provider) {
        this.provider = provider;
        this.request = new ShhRequestPayloads();
    }
    send({ method, params, format }) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return format(yield this.provider.send(method, params));
        });
    }
    getVersion() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.getVersion();
            return payload.format(yield this.send(payload));
        });
    }
    getInfo() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.getInfo();
            return payload.format(yield this.send(payload));
        });
    }
    setMaxMessageSize(size) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.setMaxMessageSize(size);
            return payload.format(yield this.send(payload));
        });
    }
    setMinPoW(pow) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.setMinPoW(pow);
            return payload.format(yield this.send(payload));
        });
    }
    markTrustedPeer(enode) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.markTrustedPeer(enode);
            return payload.format(yield this.send(payload));
        });
    }
    newKeyPair() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.newKeyPair();
            return payload.format(yield this.send(payload));
        });
    }
    addPrivateKey(privateKey) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.addPrivateKey(privateKey);
            return payload.format(yield this.send(payload));
        });
    }
    deleteKeyPair(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.deleteKeyPair(id);
            return payload.format(yield this.send(payload));
        });
    }
    hasKeyPair(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.hasKeyPair(id);
            return payload.format(yield this.send(payload));
        });
    }
    getPublicKey(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.getPublicKey(id);
            return payload.format(yield this.send(payload));
        });
    }
    getPrivateKey(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.getPrivateKey(id);
            return payload.format(yield this.send(payload));
        });
    }
    newSymKey() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.newSymKey();
            return payload.format(yield this.send(payload));
        });
    }
    addSymKey(symKey) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.addSymKey(symKey);
            return payload.format(yield this.send(payload));
        });
    }
    generateSymKeyFromPassword(password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.generateSymKeyFromPassword(password);
            return payload.format(yield this.send(payload));
        });
    }
    hasSymKey(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.hasSymKey(id);
            return payload.format(yield this.send(payload));
        });
    }
    getSymKey(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.getSymKey(id);
            return payload.format(yield this.send(payload));
        });
    }
    deleteSymKey(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.deleteSymKey(id);
            return payload.format(yield this.send(payload));
        });
    }
    newMessageFilter(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.newMessageFilter(options);
            return payload.format(yield this.send(payload));
        });
    }
    getFilterMessages(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.getFilterMessages(id);
            return payload.format(yield this.send(payload));
        });
    }
    deleteMessageFilter(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.deleteMessageFilter(id);
            return payload.format(yield this.send(payload));
        });
    }
    post(post) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.post(post);
            return payload.format(yield this.send(payload));
        });
    }
    subscribeMessages(options) {
        const subscription = new subscriptions_1.Subscription('shh', 'messages', [options], this.provider);
        subscription.on('rawdata', message => subscription.emit('data', message));
        subscription.subscribe();
        return subscription;
    }
    subscribe(type, options) {
        switch (type) {
            case 'messages':
                return this.subscribeMessages(options);
            default:
                throw new Error(`Unknown subscription type: ${type}`);
        }
    }
}
exports.Shh = Shh;
