// web3x/ens/ens.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const config_1 = require("./config");
const registry_1 = require("./registry");
const namehash_1 = require("./registry/namehash");
const net_1 = require("../net");
/**
 * Constructs a new instance of ENS
 *
 * @method ENS
 * @param {Object} eth
 * @constructor
 */
class ENS {
    constructor(eth) {
        this.eth = eth;
        this.registry = new registry_1.Registry(this);
        this.net = new net_1.Net(eth);
    }
    getRegistry() {
        return this.registry;
    }
    /**
     * @param {string} name
     * @returns {Promise<Contract>}
     */
    getResolver(name) {
        return this.registry.resolver(name);
    }
    /**
     * Returns the address record associated with a name.
     *
     * @method getAddress
     * @param {string} name
     * @param {function} callback
     * @return {eventifiedPromise}
     */
    getAddress(name) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.addr(namehash_1.namehash(name)).call();
        });
    }
    /**
     * Sets a new address
     *
     * @method setAddress
     * @param {string} name
     * @param {string} address
     * @param {Object} sendOptions
     * @param {function} callback
     * @returns {eventifiedPromise}
     */
    setAddress(name, address, sendOptions) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.setAddr(namehash_1.namehash(name), address).send(sendOptions);
        });
    }
    /**
     * Returns the public key
     *
     * @method getPubkey
     * @param {string} name
     * @param {function} callback
     * @returns {eventifiedPromise}
     */
    getPubkey(name) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.pubkey(namehash_1.namehash(name)).call();
        });
    }
    /**
     * Set the new public key
     *
     * @method setPubkey
     * @param {string} name
     * @param {string} x
     * @param {string} y
     * @param {Object} sendOptions
     * @param {function} callback
     * @returns {eventifiedPromise}
     */
    setPubkey(name, x, y, sendOptions) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.setPubkey(namehash_1.namehash(name), x, y).send(sendOptions);
        });
    }
    /**
     * Returns the content
     *
     * @method getContent
     * @param {string} name
     * @param {function} callback
     * @returns {eventifiedPromise}
     */
    getContent(name) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.content(namehash_1.namehash(name)).call();
        });
    }
    /**
     * Set the content
     *
     * @method setContent
     * @param {string} name
     * @param {string} hash
     * @param {function} callback
     * @param {Object} sendOptions
     * @returns {eventifiedPromise}
     */
    setContent(name, hash, sendOptions) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.setContent(namehash_1.namehash(name), hash).send(sendOptions);
        });
    }
    /**
     * Get the multihash
     *
     * @method getMultihash
     * @param {string} name
     * @param {function} callback
     * @returns {eventifiedPromise}
     */
    getMultihash(name) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.multihash(namehash_1.namehash(name)).call();
        });
    }
    /**
     * Set the multihash
     *
     * @method setMultihash
     * @param {string} name
     * @param {string} hash
     * @param {Object} sendOptions
     * @param {function} callback
     * @returns {eventifiedPromise}
     */
    setMultihash(name, hash, sendOptions) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resolver = yield this.registry.resolver(name);
            return yield resolver.methods.setMultihash(namehash_1.namehash(name), hash).send(sendOptions);
        });
    }
    /**
     * Checks if the current used network is synced and looks for ENS support there.
     * Throws an error if not.
     *
     * @returns {Promise<Block>}
     */
    checkNetwork() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const block = yield this.eth.getBlock('latest');
            const headAge = new Date().getTime() / 1000 - block.timestamp;
            if (headAge > 3600) {
                throw new Error('Network not synced; last block was ' + headAge + ' seconds ago');
            }
            const networkType = yield this.net.getNetworkType();
            const addr = config_1.config.addresses[networkType];
            if (typeof addr === 'undefined') {
                throw new Error('ENS is not supported on network ' + networkType);
            }
            return addr;
        });
    }
}
exports.ENS = ENS;
