// web3x/contract/tx-deploy.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const util_1 = require("util");
const promievent_1 = require("../promievent");
const abi_1 = require("./abi");
const utils_1 = require("../utils");
const formatters_1 = require("../formatters");
/**
 * returns the an object with call, send, estimate functions
 *
 * @method _createTxObject
 * @returns {Object} an object with functions to call the methods
 */
class TxDeploy {
    constructor(eth, definition, deployData, args = [], defaultOptions = {}, wallet, extraFormatters) {
        this.eth = eth;
        this.definition = definition;
        this.deployData = deployData;
        this.args = args;
        this.defaultOptions = defaultOptions;
        this.wallet = wallet;
        this.extraFormatters = extraFormatters;
        if (this.defaultOptions.from) {
            this.defaultOptions.from = utils_1.toChecksumAddress(formatters_1.inputAddressFormatter(this.defaultOptions.from));
        }
    }
    estimateGas(options = {}) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return yield this.eth.estimateGas(this.getTx(options));
        });
    }
    send(options) {
        const tx = this.getTx(options);
        if (util_1.isBoolean(this.definition.payable) && !this.definition.payable && tx.value && tx.value > 0) {
            const defer = promievent_1.promiEvent();
            return utils_1.fireError(new Error('Can not send value to non-payable contract method or constructor'), defer.eventEmitter, defer.reject);
        }
        const account = this.getAccount(tx.from);
        if (account) {
            return account.sendTransaction(tx, this.extraFormatters);
        }
        else {
            return this.eth.sendTransaction(tx, this.extraFormatters);
        }
    }
    getRequestPayload(options) {
        return this.eth.request.sendTransaction(this.getTx(options));
    }
    getAccount(address) {
        address = address || this.defaultOptions.from;
        if (this.wallet && address) {
            return this.wallet.get(address);
        }
    }
    getTx(options) {
        return {
            from: options.from ? utils_1.toChecksumAddress(formatters_1.inputAddressFormatter(options.from)) : this.defaultOptions.from,
            gasPrice: options.gasPrice || this.defaultOptions.gasPrice,
            gas: options.gas || this.defaultOptions.gas,
            value: options.value,
            data: this.encodeABI(),
        };
    }
    /**
     * Encodes an ABI for a method, including signature or the method.
     * Or when constructor encodes only the constructor parameters.
     *
     * @method encodeABI
     * @param {Mixed} args the arguments to encode
     * @param {String} the encoded ABI
     */
    encodeABI() {
        let paramsABI = abi_1.abi.encodeParameters(this.definition.inputs || [], this.args).replace('0x', '');
        return this.deployData + paramsABI;
    }
}
exports.TxDeploy = TxDeploy;
