// web3x/subscriptions/subscription.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const util_1 = require("util");
const events_1 = require("events");
class Subscription extends events_1.EventEmitter {
    constructor(type, subscription, params, provider) {
        super();
        this.type = type;
        this.subscription = subscription;
        this.params = params;
        this.provider = provider;
    }
    on(event, listener) {
        return super.on(event, listener);
    }
    once(event, listener) {
        return super.once(event, listener);
    }
    subscribe() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.id) {
                this.unsubscribe();
            }
            try {
                this.listener = params => this.notificationHandler(params);
                this.provider.on('notification', this.listener);
                this.id = yield this.provider.send(`${this.type}_subscribe`, [this.subscription, ...this.params]);
                if (!this.id) {
                    throw new Error('No result');
                }
            }
            catch (err) {
                this.unsubscribe();
                this.emit('error', err);
            }
            return this;
        });
    }
    notificationHandler(params) {
        const { subscription, result } = params;
        if (subscription !== this.id) {
            return;
        }
        if (result instanceof Error) {
            this.emit('error', result);
            return;
        }
        const resultArr = util_1.isArray(result) ? result : [result];
        resultArr.forEach(resultItem => {
            this.emit('rawdata', resultItem);
        });
    }
    unsubscribe() {
        if (this.listener) {
            this.provider.removeListener('notification', this.listener);
        }
        this.removeAllListeners();
        this.id = undefined;
        this.listener = undefined;
    }
}
exports.Subscription = Subscription;
