// web3x/personal/personal.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const personal_request_payloads_1 = require("./personal-request-payloads");
class Personal {
    constructor(provider) {
        this.provider = provider;
        this.request = new personal_request_payloads_1.PersonalRequestPayloads();
    }
    send({ method, params, format }) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return format(yield this.provider.send(method, params));
        });
    }
    getAccounts() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.getAccounts();
            return payload.format(yield this.send(payload));
        });
    }
    newAccount(password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.newAccount(password);
            return payload.format(yield this.send(payload));
        });
    }
    unlockAccount(address, password, duration) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.unlockAccount(address, password, duration);
            return payload.format(yield this.send(payload));
        });
    }
    lockAccount(address) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.lockAccount(address);
            return payload.format(yield this.send(payload));
        });
    }
    importRawKey(privateKey, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.importRawKey(privateKey, password);
            return payload.format(yield this.send(payload));
        });
    }
    sendTransaction(tx, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.sendTransaction(tx, password);
            return payload.format(yield this.send(payload));
        });
    }
    signTransaction(tx, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.signTransaction(tx, password);
            return payload.format(yield this.send(payload));
        });
    }
    sign(data, address, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.sign(data, address, password);
            return payload.format(yield this.send(payload));
        });
    }
    ecRecover(data, signedData) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const payload = this.request.ecRecover(data, signedData);
            return payload.format(yield this.send(payload));
        });
    }
}
exports.Personal = Personal;
