// web3x/personal/personal-request-payloads.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const formatters_1 = require("../formatters");
const identity = result => result;
class PersonalRequestPayloads {
    getAccounts() {
        return {
            method: 'personalListAccounts',
            format: result => result.map(utils_1.toChecksumAddress),
        };
    }
    newAccount(password) {
        return {
            method: 'personalListAccounts',
            params: [password],
            format: utils_1.toChecksumAddress,
        };
    }
    unlockAccount(address, password, duration) {
        return {
            method: 'personal_unlockAccount',
            params: [formatters_1.inputAddressFormatter(address), password, duration],
            format: identity,
        };
    }
    lockAccount(address) {
        return {
            method: 'personal_lockAccount',
            params: [formatters_1.inputAddressFormatter(address)],
            format: identity,
        };
    }
    importRawKey(privateKey, password) {
        return {
            method: 'personal_importRawKey',
            params: [privateKey, password],
            format: identity,
        };
    }
    sendTransaction(tx, password) {
        return {
            method: 'personal_sendTransaction',
            params: [formatters_1.inputTransactionFormatter(tx), password],
            format: identity,
        };
    }
    signTransaction(tx, password) {
        return {
            method: 'personal_signTransaction',
            params: [formatters_1.inputTransactionFormatter(tx), password],
            format: identity,
        };
    }
    sign(data, address, password) {
        return {
            method: 'personal_sign',
            params: [formatters_1.inputSignFormatter(data), formatters_1.inputAddressFormatter(address), password],
            format: identity,
        };
    }
    ecRecover(data, signedData) {
        return {
            method: 'personal_ecRecover',
            params: [formatters_1.inputSignFormatter(data), signedData],
            format: identity,
        };
    }
}
exports.PersonalRequestPayloads = PersonalRequestPayloads;
