// web3x/wallet/wallet.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const util_1 = require("util");
const account_1 = require("../account");
const encryption_1 = require("../utils/encryption");
class Wallet {
    constructor() {
        this.length = 0;
        this.accounts = [];
    }
    static fromKeystores(encryptedWallet, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const wallet = new Wallet();
            yield wallet.decrypt(encryptedWallet, password);
            return wallet;
        });
    }
    static fromLocalStorage(password, keyName = this.defaultKeyName) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!localStorage) {
                return new Wallet();
            }
            const keystoreStr = localStorage.getItem(keyName);
            if (!keystoreStr) {
                return new Wallet();
            }
            try {
                return Wallet.fromKeystores(JSON.parse(keystoreStr), password);
            }
            catch (e) {
                return new Wallet();
            }
        });
    }
    findSafeIndex(pointer = 0) {
        while (this.accounts[pointer]) {
            ++pointer;
        }
        return pointer;
    }
    currentIndexes() {
        const keys = Object.keys(this.accounts);
        return keys.map(key => +key);
    }
    create(numberOfAccounts, entropy) {
        for (var i = 0; i < numberOfAccounts; ++i) {
            this.add(account_1.Account.create(entropy).privateKey);
        }
        return this.accounts;
    }
    get(addressOrIndex) {
        if (util_1.isString(addressOrIndex)) {
            return this.accounts.find(a => a && a.address.toLowerCase() === addressOrIndex.toLowerCase());
        }
        return this.accounts[addressOrIndex];
    }
    indexOf(addressOrIndex) {
        if (util_1.isString(addressOrIndex)) {
            return this.accounts.findIndex(a => a.address.toLowerCase() === addressOrIndex.toLowerCase());
        }
        return addressOrIndex;
    }
    add(accountOrKey) {
        if (Buffer.isBuffer(accountOrKey)) {
            accountOrKey = account_1.Account.fromPrivate(accountOrKey);
        }
        else {
            accountOrKey = account_1.Account.fromPrivate(accountOrKey.privateKey);
        }
        const existing = this.get(accountOrKey.address);
        if (existing) {
            return existing;
        }
        const index = this.findSafeIndex();
        this.accounts[index] = accountOrKey;
        this.length++;
        return accountOrKey;
    }
    remove(addressOrIndex) {
        const index = this.indexOf(addressOrIndex);
        if (index == -1) {
            return false;
        }
        this.accounts[index].privateKey = Buffer.of();
        delete this.accounts[index];
        this.length--;
        return true;
    }
    clear() {
        var indexes = this.currentIndexes();
        indexes.forEach(index => {
            this.remove(index);
        });
    }
    encrypt(password, options) {
        return Promise.all(this.currentIndexes().map(index => this.accounts[index].encrypt(password, options)));
    }
    decrypt(encryptedWallet, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const decrypted = yield Promise.all(encryptedWallet.map(keystore => encryption_1.decrypt(keystore, password)));
            decrypted.forEach(account => {
                if (!account) {
                    throw new Error("Couldn't decrypt accounts. Password wrong?");
                }
                this.add(account);
            });
            return this.accounts;
        });
    }
    save(password, keyName = Wallet.defaultKeyName) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!localStorage) {
                return false;
            }
            localStorage.setItem(keyName, JSON.stringify(yield this.encrypt(password)));
            return true;
        });
    }
}
Wallet.defaultKeyName = 'web3js_wallet';
exports.Wallet = Wallet;
