// web3x/eth-lib/account.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bytes_1 = tslib_1.__importDefault(require("./bytes"));
const nat_1 = tslib_1.__importDefault(require("./nat"));
const elliptic_1 = tslib_1.__importDefault(require("elliptic"));
const hash_1 = require("./hash");
const secp256k1 = new elliptic_1.default.ec('secp256k1');
exports.create = (entropy) => {
    const innerHex = hash_1.keccak256(bytes_1.default.concat(bytes_1.default.random(32), '0x' + entropy.toString('hex') || bytes_1.default.random(32)));
    const middleHex = bytes_1.default.concat(bytes_1.default.concat(bytes_1.default.random(32), innerHex), bytes_1.default.random(32));
    const outerHex = hash_1.keccak256(middleHex);
    return exports.fromPrivate(Buffer.from(outerHex.slice(2), 'hex'));
};
exports.toChecksum = address => {
    const addressHash = hash_1.keccak256s(address.slice(2));
    let checksumAddress = '0x';
    for (let i = 0; i < 40; i++)
        checksumAddress += parseInt(addressHash[i + 2], 16) > 7 ? address[i + 2].toUpperCase() : address[i + 2];
    return checksumAddress;
};
exports.fromPrivate = (privateKey) => {
    const ecKey = secp256k1.keyFromPrivate(privateKey);
    const publicKey = '0x' + ecKey.getPublic(false, 'hex').slice(2);
    const publicHash = hash_1.keccak256(publicKey);
    const address = exports.toChecksum('0x' + publicHash.slice(-40));
    return {
        address,
        privateKey,
        publicKey,
    };
};
exports.encodeSignature = ([v, r, s]) => bytes_1.default.flatten([bytes_1.default.pad(32, r), bytes_1.default.pad(32, s), v]);
exports.decodeSignature = (hex) => [
    bytes_1.default.slice(64, bytes_1.default.length(hex), hex),
    bytes_1.default.slice(0, 32, hex),
    bytes_1.default.slice(32, 64, hex),
];
exports.makeSigner = addToV => (hash, privateKey) => {
    const signature = secp256k1.keyFromPrivate(privateKey).sign(Buffer.from(hash.slice(2), 'hex'), { canonical: true });
    return exports.encodeSignature([
        nat_1.default.fromString(bytes_1.default.fromNumber(addToV + signature.recoveryParam)),
        bytes_1.default.pad(32, bytes_1.default.fromNat('0x' + signature.r.toString(16))),
        bytes_1.default.pad(32, bytes_1.default.fromNat('0x' + signature.s.toString(16))),
    ]);
};
exports.sign = exports.makeSigner(27); // v=27|28 instead of 0|1...
exports.recover = (hash, signature) => {
    const vals = exports.decodeSignature(signature);
    const vrs = { v: bytes_1.default.toNumber(vals[0]), r: vals[1].slice(2), s: vals[2].slice(2) };
    const ecPublicKey = secp256k1.recoverPubKey(Buffer.from(hash.slice(2), 'hex'), vrs, vrs.v < 2 ? vrs.v : 1 - (vrs.v % 2)); // because odd vals mean v=0... sadly that means v=0 means v=1... I hate that
    const publicKey = '0x' + ecPublicKey.encode('hex', false).slice(2);
    const publicHash = hash_1.keccak256(publicKey);
    const address = exports.toChecksum('0x' + publicHash.slice(-40));
    return address;
};
exports.default = { create: exports.create, toChecksum: exports.toChecksum, fromPrivate: exports.fromPrivate, sign: exports.sign, makeSigner: exports.makeSigner, recover: exports.recover, encodeSignature: exports.encodeSignature, decodeSignature: exports.decodeSignature };
