// web3x/utils/encryption.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const browserify_aes_1 = tslib_1.__importDefault(require("browserify-aes"));
const randombytes_1 = tslib_1.__importDefault(require("randombytes"));
const util_1 = require("util");
const uuid_1 = tslib_1.__importDefault(require("uuid"));
const _1 = require(".");
function decrypt(v3Keystore, password, nonStrict = false) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (!util_1.isString(password)) {
            throw new Error('No password given.');
        }
        var json = !util_1.isString(v3Keystore) ? v3Keystore : JSON.parse(nonStrict ? v3Keystore.toLowerCase() : v3Keystore);
        if (json.version !== 3) {
            throw new Error('Not a valid V3 wallet');
        }
        let derivedKey;
        if (json.crypto.kdf === 'scrypt') {
            const { n, r, p, dklen, salt } = json.crypto.kdfparams;
            derivedKey = yield _1.scrypt(Buffer.from(password), Buffer.from(salt, 'hex'), n, r, p, dklen);
        }
        else if (json.crypto.kdf === 'pbkdf2') {
            const { prf, c, dklen, salt } = json.crypto.kdfparams;
            if (prf !== 'hmac-sha256') {
                throw new Error('Unsupported parameters to PBKDF2');
            }
            derivedKey = yield _1.pbkdf2(Buffer.from(password), Buffer.from(salt, 'hex'), c, dklen);
        }
        else {
            throw new Error('Unsupported key derivation scheme');
        }
        const ciphertext = Buffer.from(json.crypto.ciphertext, 'hex');
        const mac = _1.sha3(Buffer.concat([derivedKey.slice(16, 32), ciphertext])).replace('0x', '');
        if (mac !== json.crypto.mac) {
            throw new Error('Key derivation failed - possibly wrong password');
        }
        const iv = Buffer.from(json.crypto.cipherparams.iv, 'hex');
        const aesKey = derivedKey.slice(0, 16);
        const decipher = browserify_aes_1.default.createDecipheriv(json.crypto.cipher, aesKey, iv);
        return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    });
}
exports.decrypt = decrypt;
function encrypt(privateKey, address, password, options = {}) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const cipherAlgo = options.cipher || 'aes-128-ctr';
        const salt = options.salt ? Buffer.from(options.salt, 'hex') : randombytes_1.default(32);
        const iv = options.iv ? Buffer.from(options.iv, 'hex') : randombytes_1.default(16);
        const kdf = options.kdf || 'scrypt';
        const id = options.id || uuid_1.default.v4({ random: options.uuid || randombytes_1.default(16) });
        if (cipherAlgo !== 'aes-128-ctr') {
            throw new Error('Unsupported cipher');
        }
        let derivedKey;
        let kdfparams;
        if (kdf === 'pbkdf2') {
            const { c = 262144, dklen = 32 } = options;
            derivedKey = yield _1.pbkdf2(Buffer.from(password), salt, c, dklen);
            kdfparams = { c, dklen, prf: 'hmac-sha256', salt: salt.toString('hex') };
        }
        else if (kdf === 'scrypt') {
            const { n = 8192, r = 8, p = 1, dklen = 32 } = options;
            derivedKey = yield _1.scrypt(Buffer.from(password), salt, n, r, p, dklen);
            kdfparams = { n, r, p, dklen, salt: salt.toString('hex') };
        }
        else {
            throw new Error('Unsupported kdf');
        }
        const aesKey = derivedKey.slice(0, 16);
        const cipher = browserify_aes_1.default.createCipheriv(cipherAlgo, aesKey, iv);
        if (!cipher) {
            throw new Error('Unsupported cipher');
        }
        const ciphertext = Buffer.concat([cipher.update(privateKey), cipher.final()]);
        const mac = _1.sha3(Buffer.concat([derivedKey.slice(16, 32), ciphertext])).replace('0x', '');
        return {
            version: 3,
            id,
            address: address.toLowerCase().replace('0x', ''),
            crypto: {
                ciphertext: ciphertext.toString('hex'),
                cipherparams: {
                    iv: iv.toString('hex'),
                },
                cipher: 'aes-128-ctr',
                kdf: kdf,
                kdfparams: kdfparams,
                mac: mac.toString(),
            },
        };
    });
}
exports.encrypt = encrypt;
