// web3x/promievent/index.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
/**
 * This function generates a defer promise and adds eventEmitter functionality to it
 *
 * @method eventifiedPromise
 */
class PromiEvent {
    constructor(promise, emitter) {
        this.promise = promise;
        this.emitter = emitter;
    }
    then(onfulfilled, onrejected) {
        return new PromiEvent(this.promise.then(onfulfilled, onrejected), this.emitter);
    }
    catch(onrejected) {
        return new PromiEvent(this.promise.catch(onrejected), this.emitter);
    }
    finally(onfinally) {
        const promise = this.promise;
        if (promise.finally) {
            return new PromiEvent(promise.finally(onfinally), this.emitter);
        }
        else {
            throw new Error('Underlying Promise library does not support finally.');
        }
    }
    addListener(event, listener) {
        this.emitter.addListener(event, listener);
        return this;
    }
    on(event, listener) {
        this.emitter.on(event, listener);
        return this;
    }
    once(event, listener) {
        this.emitter.once(event, listener);
        return this;
    }
    prependListener(event, listener) {
        this.emitter.prependListener(event, listener);
        return this;
    }
    prependOnceListener(event, listener) {
        this.emitter.prependOnceListener(event, listener);
        return this;
    }
    removeListener(event, listener) {
        this.emitter.removeListener(event, listener);
        return this;
    }
    off(event, listener) {
        this.emitter.off(event, listener);
        return this;
    }
    removeAllListeners(event) {
        this.emitter.removeAllListeners(event);
        return this;
    }
    setMaxListeners(n) {
        this.emitter.setMaxListeners(n);
        return this;
    }
    getMaxListeners() {
        return this.emitter.getMaxListeners();
    }
    listeners(event) {
        return this.emitter.listeners(event);
    }
    rawListeners(event) {
        return this.emitter.rawListeners(event);
    }
    emit(event, ...args) {
        return this.emitter.emit(event, ...args);
    }
    eventNames() {
        return this.emitter.eventNames();
    }
    listenerCount(type) {
        return this.emitter.listenerCount(type);
    }
}
exports.PromiEvent = PromiEvent;
function promiEvent(justPromise = false) {
    let resolve, reject;
    const promise = new Promise((res, rej) => {
        resolve = res;
        reject = rej;
    });
    if (justPromise) {
        return { resolve, reject, eventEmitter: promise };
    }
    const emitter = new events_1.EventEmitter();
    // Required to suppress default behaviour of throwing on emitting an error.
    emitter.on('error', () => { });
    const pe = new PromiEvent(promise, emitter);
    return {
        resolve,
        reject,
        eventEmitter: pe,
    };
}
exports.promiEvent = promiEvent;
function resolvedPromiEvent(value) {
    const pe = promiEvent(true);
    pe.resolve(value);
    return pe.eventEmitter;
}
exports.resolvedPromiEvent = resolvedPromiEvent;
