// web3x/account/account.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bip39_1 = tslib_1.__importDefault(require("bip39"));
const hdkey_1 = tslib_1.__importDefault(require("hdkey"));
const account_1 = require("../eth-lib/account");
const utils_1 = require("../utils");
const sign_1 = require("../utils/sign");
const sign_transaction_1 = require("./sign-transaction");
const promievent_1 = require("../promievent");
class Account {
    constructor(address, privateKey, publicKey) {
        this.address = address;
        this.privateKey = privateKey;
        this.publicKey = publicKey;
    }
    static create(entropy = utils_1.randomHex(32)) {
        const { privateKey, address, publicKey } = account_1.create(entropy);
        return new Account(address, privateKey, publicKey);
    }
    static fromPrivate(privateKey) {
        const { address, publicKey } = account_1.fromPrivate(privateKey);
        return new Account(address, privateKey, publicKey);
    }
    static createFromMnemonicAndPath(mnemonic, derivationPath) {
        const seed = bip39_1.default.mnemonicToSeed(mnemonic);
        return Account.createFromSeedAndPath(seed, derivationPath);
    }
    static createFromSeedAndPath(seed, derivationPath) {
        const root = hdkey_1.default.fromMasterSeed(seed);
        const addrNode = root.derive(derivationPath);
        const privateKey = addrNode.privateKey;
        return Account.fromPrivate(privateKey);
    }
    static fromKeystore(v3Keystore, password, nonStrict = false) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return Account.fromPrivate(yield utils_1.decrypt(v3Keystore, password, nonStrict));
        });
    }
    sendTransaction(tx, eth, extraformatters) {
        const defer = promievent_1.promiEvent();
        this.signTransaction(tx, eth)
            .then(signedTx => {
            eth.sendSignedTransaction(signedTx.rawTransaction, extraformatters, defer);
        })
            .catch(err => {
            utils_1.fireError(err, defer.eventEmitter, defer.reject);
        });
        return defer.eventEmitter;
    }
    signTransaction(tx, eth) {
        return sign_transaction_1.signTransaction(tx, this.privateKey, eth);
    }
    sign(data) {
        return sign_1.sign(data, this.privateKey);
    }
    encrypt(password, options) {
        return utils_1.encrypt(this.privateKey, this.address, password, options);
    }
}
exports.Account = Account;
