// web3x/account/sign-transaction.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const utils_1 = require("../utils");
const rlp_1 = tslib_1.__importDefault(require("../eth-lib/rlp"));
const bytes_1 = tslib_1.__importDefault(require("../eth-lib/bytes"));
const hash_1 = tslib_1.__importDefault(require("../eth-lib/hash"));
const nat_1 = tslib_1.__importDefault(require("../eth-lib/nat"));
const account_1 = tslib_1.__importDefault(require("../eth-lib/account"));
const formatters_1 = require("../formatters");
function signTransaction(tx, privateKey, eth) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (!tx.gas) {
            throw new Error('gas is missing or 0');
        }
        // Resolve immediately if nonce, chainId and price are provided
        if (tx.nonce !== undefined && tx.chainId !== undefined && tx.gasPrice !== undefined) {
            return sign(tx, privateKey);
        }
        // Otherwise, get the missing info from the Ethereum Node
        const promises = [
            isNot(tx.chainId) ? eth.getId() : Promise.resolve(tx.chainId),
            isNot(tx.gasPrice) ? eth.getGasPrice() : Promise.resolve(tx.gasPrice),
            isNot(tx.nonce) ? eth.getTransactionCount(account_1.default.fromPrivate(privateKey).address) : Promise.resolve(tx.nonce),
        ];
        const [chainId, gasPrice, nonce] = yield Promise.all(promises);
        if (isNot(chainId) || isNot(gasPrice) || isNot(nonce)) {
            throw new Error('One of the values chainId, gasPrice, or nonce could not be fetched');
        }
        return sign(Object.assign({}, tx, { chainId, gasPrice, nonce }), privateKey);
    });
}
exports.signTransaction = signTransaction;
function recoverTransaction(rawTx) {
    var values = rlp_1.default.decode(rawTx);
    var signature = account_1.default.encodeSignature(values.slice(6, 9));
    var recovery = bytes_1.default.toNumber(values[6]);
    var extraData = recovery < 35 ? [] : [bytes_1.default.fromNumber((recovery - 35) >> 1), '0x', '0x'];
    var signingData = values.slice(0, 6).concat(extraData);
    var signingDataHex = rlp_1.default.encode(signingData);
    return account_1.default.recover(hash_1.default.keccak256(signingDataHex), signature);
}
exports.recoverTransaction = recoverTransaction;
function sign(tx, privateKey) {
    if (tx.nonce < 0 || tx.gas < 0 || tx.gasPrice < 0 || tx.chainId < 0) {
        throw new Error('gas, gasPrice, nonce or chainId is lower than 0');
    }
    tx = formatters_1.inputCallFormatter(tx);
    const transaction = tx;
    transaction.to = tx.to || '0x';
    transaction.data = tx.data || '0x';
    transaction.value = tx.value || '0x';
    transaction.chainId = utils_1.numberToHex(tx.chainId);
    const rlpEncoded = rlp_1.default.encode([
        bytes_1.default.fromNat(transaction.nonce),
        bytes_1.default.fromNat(transaction.gasPrice),
        bytes_1.default.fromNat(transaction.gas),
        transaction.to.toLowerCase(),
        bytes_1.default.fromNat(transaction.value),
        transaction.data,
        bytes_1.default.fromNat(transaction.chainId || '0x1'),
        '0x',
        '0x',
    ]);
    const messageHash = hash_1.default.keccak256(rlpEncoded);
    const signature = account_1.default.makeSigner(nat_1.default.toNumber(transaction.chainId || '0x1') * 2 + 35)(hash_1.default.keccak256(rlpEncoded), privateKey);
    const rawTx = rlp_1.default.decode(rlpEncoded)
        .slice(0, 6)
        .concat(account_1.default.decodeSignature(signature));
    rawTx[6] = makeEven(trimLeadingZero(rawTx[6]));
    rawTx[7] = makeEven(trimLeadingZero(rawTx[7]));
    rawTx[8] = makeEven(trimLeadingZero(rawTx[8]));
    const rawTransaction = rlp_1.default.encode(rawTx);
    const values = rlp_1.default.decode(rawTransaction);
    return {
        messageHash,
        v: trimLeadingZero(values[6]),
        r: trimLeadingZero(values[7]),
        s: trimLeadingZero(values[8]),
        rawTransaction,
    };
}
function isNot(value) {
    return value === undefined || value === null;
}
function trimLeadingZero(hex) {
    while (hex && hex.startsWith('0x0')) {
        hex = '0x' + hex.slice(3);
    }
    return hex;
}
function makeEven(hex) {
    if (hex.length % 2 === 1) {
        hex = hex.replace('0x', '0x0');
    }
    return hex;
}
